### ISSUE TITLE
DateTime in Indonesian/Jakarta timezone not displaying correctly for Jan 1, 1964

### ISSUE THREAD

**Issue by shihaohong** [dart-lang/sdk#44779]
I'm not sure where this exactly goes, but the docs seem to indicate that DateTime is a dart:core feature.

In the Flutter DatePicker widget, we make use of the DateTime API to get the dates for the calendar widget population. It seems to work in all other timezones, except for Indonesia/Jakarta (GMT+7), for a very particular date (Jan 1, 1964). A reproducible sample is to set the native device/platform's timezone to Indonesia/Jakarta, and then:

```dart
void main() {
  print(DateTime(1964, 1)); // 1963-12-31 23:30:00.000
}
```

Basically, instead of yielding 1964-01-01 00:00:00.000, it results in a time 30 minutes prior.

Related Flutter issue: https://github.com/flutter/flutter/issues/74303

**Comment by mraleph** [comment]
According to the time zone data they changed from GMT+7:30 to GMT+7 on that particular date. That's why it is 30m off. 

/cc @lrhn to confirm.

**Comment by lrhn** [comment]
That could very well be the reason, yes.

If they changed their time zone from GMT+7:30 to GMT+7 at the end of 1963-12-31(GMT+7.30), then the moment 1964-01-01T00:00:00 in GMT+7.30 is also 1963-12-31T23:30:00 in GMT+7.

I do think it's a little weird that we don't simply choose 1964-01-01T00:00:00 GMT+7 as the local time (which is half an hour later). I don't actually know how we create local-time dates, it's based on the time zone information provided by the underlying platform.

**Comment by shihaohong** [comment]
Would it be possible to change it to choose `1964-01-01T00:00:00 GMT+7` over `1963-12-31T23:30:00 in GMT+7`?

**Comment by lrhn** [comment]
We are trying to match the JavaScript `Date` constructor behavior here, so it's not clear how much we can afford to deviate.

It's *probably* possible to do something different (although I can't guarantee that there won't be some time zone transition which makes it completely undecidable what to do). Maybe can do binary search for a time where the time-zone offset precisely matches the millisecond difference from UTC time.
It would probably add some overhead to every other `DateTime` construction, though.

**Comment by lrhn** [comment]
I have changed the computation to work better with non-1-hour offsets. It should be part of the next Dart stable release, and is already on the dev channel.

**Comment by lrhn** [comment]
No further plans here.
